src/hotspot/share/opto/predicates.cpp

```diff
@@ -170,32 +170,37 @@ bool TemplateAssertionPredicate::is_predicate(Node* node) {
   return if_node->in(1)->is_OpaqueTemplateAssertionPredicate();
 }
 
-// Clone this Template Assertion Predicate and replace the OpaqueLoopInitNode with the provided 'new_opaque_init' node.
-IfTrueNode* TemplateAssertionPredicate::clone(Node* new_control, PhaseIdealLoop* phase) const {
+// Clone this Template Assertion Predicate without modifying any OpaqueLoop*Node inputs.
+TemplateAssertionPredicate TemplateAssertionPredicate::clone(Node* new_control, PhaseIdealLoop* phase) const {
   DEBUG_ONLY(verify();)
   TemplateAssertionExpression template_assertion_expression(opaque_node());
   OpaqueTemplateAssertionPredicateNode* new_opaque_node = template_assertion_expression.clone(new_control, phase);
   AssertionPredicateIfCreator assertion_predicate_if_creator(phase);
   IfTrueNode* success_proj = assertion_predicate_if_creator.create_for_template(new_control, _if_node->Opcode(),
                                                                                 new_opaque_node,
                                                                                 _if_node->assertion_predicate_type());
-  DEBUG_ONLY(TemplateAssertionPredicate::verify(success_proj);)
-  return success_proj;
+  TemplateAssertionPredicate cloned_template_assertion_predicate(success_proj);
+  DEBUG_ONLY(cloned_template_assertion_predicate.verify();)
+  return cloned_template_assertion_predicate;
 }
 
-// Clone this Template Assertion Predicate and replace the OpaqueLoopInitNode with the provided 'new_opaque_init' node.
-IfTrueNode* TemplateAssertionPredicate::clone_and_replace_init(Node* new_control, OpaqueLoopInitNode* new_opaque_init,
-                                                               PhaseIdealLoop* phase) const {
+// Clone this Template Assertion Predicate and replace the input of the OpaqueLoopInitNode with 'new_opaque_input'.
```

Comment:
Looks like you now also create a `new OpaqueLoopInitNode`, so that is slightly inaccurate, right?